### DISCUSSION TITLE
Add support for DoH/DoT (NextDNS DNS URL as custom DNS server)

### DISCUSSION THREAD

**Discussion by kolya182** [mullvad/mullvadvpn-app#9121] [Feature requests & Ideas]
Currently no way to use custom URL provided by @nextdns as a custom server DNS on macOS.

<img width="321" alt="Screen Shot 2022-06-22 at 12 08 58 PM" src="https://user-images.githubusercontent.com/3277207/175117203-57993f16-d1a8-4e56-8179-8a76455236e4.png">

<img width="314" alt="Screen Shot 2022-06-22 at 12 05 32 PM" src="https://user-images.githubusercontent.com/3277207/175116643-c5d788c2-a7b2-443c-adf1-00ec10aacf65.png">

**Comment by Stetsed** [comment]
While this would be a cool feature, this isn't "just" implementing it for NextDNS as this would require including a DNS-Over-HTTPS client within the Mullvad VPN app. I do however think that it would be a cool feature.

**Comment by mcmurry-1** [comment]
You could just use your **IPv6 endpoints**, they include your configuration code and work perfectly in Mullvad's macOS & iOS apps. 

![custom dns](https://user-images.githubusercontent.com/100723288/183738838-b8e0236c-bbfa-4774-9194-3f873ebe67a4.png)

**Reply by hozza** [reply]
DOH/DOT provide some authentication and security benefits that DNS53 (over IPv6) can't provide.

**Comment by solzard** [comment]
IPv6 endpoints works fine; however, DoT or DoH is required to report a device name in NextDNS logs. 
Android natively supports DoH. macOS / Linux works nicely with a `127.0.0.1` trick to fallback to NextDNS CLI. 
AFAIK, only iOS / iPadOS has no workarounds to use both Mullvad VPN and NextDNS and send logs with a device name to NextDNS.

**Comment by kolya182** [comment]
> You could just use your **IPv6 endpoints**, they include your configuration code and work perfectly in Mullvad's macOS & iOS apps.
> 

Something not quite right when using IPv6 endpoint
https://user-images.githubusercontent.com/3277207/184512083-fc3c7fec-93ce-45a9-9092-94ec6f80f397.mov

**Comment by solzard** [comment]
@kolya182 Is IPv6 supported in your environment or by ISP? 
Hotels, public WiFi, or perhaps small numbers of ISPs sometimes doesn't support IPv6.
You can check IPv6 connectivity on CLI or online (e.g. https://ipv6-test.com/). 

If IPv6 is not supported, your another solution is to use NextDNS CLI which uses DoH (i.e. https://dns.nextdns.io/xxxxxx style) internally. 
This setup doesn't get affected by whether a network supports IPv6 or not.
The commands below worked in my environment (macOS Monterey, MBP14 2021).

```sh
# Install homebrew if not yet installed.
! command -v brew &>/dev/null && /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)" 

# Install NextDNS CLI
brew install nextdns/tap/nextdns

# Configure NextDNS
# Replace <NEXTDNS_CONFIG_ID> with your own ID.
sudo "$(which nextdns)" install -config <NEXTDNS_CONFIG_ID> -report-client-info -auto-activate

# Let Mullvad VPN refer to NextDNS service
mullvad dns set custom 127.0.0.1
```

References:
[The Missing Package Manager for macOS (or Linux) — Homebrew](https://brew.sh/)
[MacOS · nextdns/nextdns Wiki](https://github.com/nextdns/nextdns/wiki/MacOS)

**Comment by steel-steven-87** [comment]
I would love to get this feature implemented on Linux, too.

**Comment by teksignal** [comment]
Any updates on this one? Would really love to have this feature on iOS

**Reply by ghost** [reply]
> Any updates on this one? Would really love to have this feature on iOS

it is never getting added

**Reply by kolya182** [reply]
> > Any updates on this one? Would really love to have this feature on iOS
> 
> 
> 
> it is never getting added

why?

**Reply by ghost** [reply]
> > > Any updates on this one? Would really love to have this feature on iOS
> > 
> > 
> > it is never getting added
> 
> why?

because, mullvad is too busy focusing and spending money on billboards/advertisements

**Comment by hozza** [comment]
I came here from issue #5192 as this is listed as a duplicate. The issue talks of adding generic DOH support rather than NextDNS DOH specifically etc.

In the ticket it was asked why this is wanted. Essentially my usecase is self hosting. I need to be able to provide custom (and not public) DNS records to access services. A custom DNS server is needed. 

It's much eaaiser to implement a DOH sever securely rather than DNS53 (old school unauthenticated DNS lookups from an IP). DoH can allow me to do custom filtering like NextDNS but also split horizon records for homelab/private services over HTTPS with authentication tokens in the DOH URL. It also prevents the DNS records being altered between mulvad exit and DNS server due to auth and TLS. 

Using a standard DNS53 IP via the app custom DNS option means I can't authenticate to my DNS server, or ensure its actually my server responding.

This is the same as NextDNS but in this case I'm my own DNS providers.

The only work around is to use Tailscale as my mobile VPN (with on network DNS server) and then use Mullvad as the eventual exit but I'd rather not use Tailscale at all.
